sys/linux/socket_netlink_route_sched.txt

```diff
@@ -95,6 +135,136 @@ q_cbs_options [
+	TCA_CODEL_LIMIT		nlattr[TCA_CODEL_LIMIT, int32]
+	TCA_CODEL_INTERVAL	nlattr[TCA_CODEL_INTERVAL, int32]
+	TCA_CODEL_ECN		nlattr[TCA_CODEL_ECN, int32[0:1]]
+	TCA_CODEL_CE_THRESHOLD	nlattr[TCA_CODEL_CE_THRESHOLD, int32]
+] [varlen]
+
+q_dsmark_options [
+#TODO: hweight32(indices) should equal to 1
+	TCA_DSMARK_INDICES		nlattr[TCA_DSMARK_INDICES, int16[0:32]]
+	TCA_DSMARK_DEFAULT_INDEX	nlattr[TCA_DSMARK_DEFAULT_INDEX, int16]
+	TCA_DSMARK_SET_TC_INDEX		nlattr[TCA_DSMARK_SET_TC_INDEX, void]
+] [varlen]
+
+q_fq_options [
+	TCA_FQ_PLIMIT			nlattr[TCA_FQ_PLIMIT, int32]
+	TCA_FQ_FLOW_PLIMIT		nlattr[TCA_FQ_FLOW_PLIMIT, int32]
+	TCA_FQ_QUANTUM			nlattr[TCA_FQ_QUANTUM, int32]
+	TCA_FQ_INITIAL_QUANTUM		nlattr[TCA_FQ_INITIAL_QUANTUM, int32]
+	TCA_FQ_RATE_ENABLE		nlattr[TCA_FQ_RATE_ENABLE, int32[0:1]]
+	TCA_FQ_FLOW_DEFAULT_RATE	nlattr[TCA_FQ_FLOW_DEFAULT_RATE, int32]
+	TCA_FQ_FLOW_MAX_RATE		nlattr[TCA_FQ_FLOW_MAX_RATE, int32]
+	TCA_FQ_BUCKETS_LOG		nlattr[TCA_FQ_BUCKETS_LOG, int32[0:32]]
+	TCA_FQ_FLOW_REFILL_DELAY	nlattr[TCA_FQ_FLOW_REFILL_DELAY, int32]
+	TCA_FQ_ORPHAN_MASK		nlattr[TCA_FQ_ORPHAN_MASK, int32]
+	TCA_FQ_LOW_RATE_THRESHOLD	nlattr[TCA_FQ_LOW_RATE_THRESHOLD, int32]
+] [varlen]
+
+q_fq_codel_options [
+	TCA_FQ_CODEL_TARGET		nlattr[TCA_FQ_CODEL_TARGET, int32]
+	TCA_FQ_CODEL_LIMIT		nlattr[TCA_FQ_CODEL_LIMIT, int32]
+	TCA_FQ_CODEL_INTERVAL		nlattr[TCA_FQ_CODEL_INTERVAL, int32]
+	TCA_FQ_CODEL_ECN		nlattr[TCA_FQ_CODEL_ECN, int32[0:1]]
+	TCA_FQ_CODEL_FLOWS		nlattr[TCA_FQ_CODEL_FLOWS, int32]
+	TCA_FQ_CODEL_QUANTUM		nlattr[TCA_FQ_CODEL_QUANTUM, int32]
+	TCA_FQ_CODEL_CE_THRESHOLD	nlattr[TCA_FQ_CODEL_CE_THRESHOLD, int32]
+	TCA_FQ_CODEL_DROP_BATCH_SIZE	nlattr[TCA_FQ_CODEL_DROP_BATCH_SIZE, int32]
+	TCA_FQ_CODEL_MEMORY_LIMIT	nlattr[TCA_FQ_CODEL_MEMORY_LIMIT, int32]
+] [varlen]
+
+# TODO: we should not have TCA_GRED_PARMS and TCA_GRED_STAB when do init
+q_gred_options [
+	TCA_GRED_PARMS	nlattr[TCA_GRED_PARMS, tc_gred_qopt]
+	TCA_GRED_STAB	nlattr[TCA_GRED_STAB, array[int8, 256]]
+	TCA_GRED_DPS	nlattr[TCA_GRED_DPS, tc_gred_sopt]
+	TCA_GRED_MAX_P	nlattr[TCA_GRED_MAX_P, int32]
+	TCA_GRED_LIMIT	nlattr[TCA_GRED_LIMIT, int32]
+] [varlen]
+
+q_hhf_options [
+	TCA_HHF_BACKLOG_LIMIT	nlattr[TCA_HHF_BACKLOG_LIMIT, int32]
+	TCA_HHF_QUANTUM		nlattr[TCA_HHF_QUANTUM, int32]
+	TCA_HHF_HH_FLOWS_LIMIT	nlattr[TCA_HHF_HH_FLOWS_LIMIT, int32]
+	TCA_HHF_RESET_TIMEOUT	nlattr[TCA_HHF_RESET_TIMEOUT, int32]
+	TCA_HHF_ADMIT_BYTES	nlattr[TCA_HHF_ADMIT_BYTES, int32]
+	TCA_HHF_EVICT_TIMEOUT	nlattr[TCA_HHF_EVICT_TIMEOUT, int32]
+	TCA_HHF_NON_HH_WEIGHT	nlattr[TCA_HHF_NON_HH_WEIGHT, int32]
+] [varlen]
+
+q_htb_options [
+	TCA_HTB_INIT		nlattr[TCA_HTB_INIT, tc_htb_glob]
+	TCA_HTB_DIRECT_QLEN	nlattr[TCA_HTB_DIRECT_QLEN, int32]
+] [varlen]
+
+q_mqprio_options [
+	TCA_MQPRIO_MODE		nlattr[TCA_MQPRIO_MODE, flags[tc_mqprio_modes, int32]]
+	TCA_MQPRIO_SHAPER	nlattr[TCA_MQPRIO_SHAPER, flags[tc_mqprio_shapers, int32]]
+	TCA_MQPRIO_MIN_RATE64	array[nlattr[TCA_MQPRIO_MIN_RATE64, int64], 16]
```

Comment:
Looking at the code, it seems that we need 1 TCA_MQPRIO_MIN_RATE64 attribute which contains N nested TCA_MQPRIO_MIN_RATE64 attributes. I.e.:
```
nlattr[TCA_MQPRIO_MIN_RATE64, array[nlattr[TCA_MQPRIO_MIN_RATE64, int64], 16]]
```

Reply:
If kernel ignores it, I don't see any reason to add it.
And I don't remember we ever added NLA_F_NESTED flags on other nested attributes.